File: app/src/main/java/social/entourage/android/sidemenu/SideMenuFragment.java
package social.entourage.android.sidemenu;


import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.iid.FirebaseInstanceId;
import com.squareup.otto.Subscribe;
import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;
import social.entourage.android.BuildConfig;
import social.entourage.android.MainActivity;
import social.entourage.android.EntourageApplication;
import social.entourage.android.R;
import social.entourage.android.api.model.Partner;
import social.entourage.android.api.model.User;
import social.entourage.android.api.model.map.Entourage;
import social.entourage.android.api.tape.Events;
import social.entourage.android.tools.BusProvider;
import social.entourage.android.tools.CropCircleTransformation;
import social.entourage.android.view.EntourageSnackbar;
import social.entourage.android.view.PartnerLogoImageView;

/**
 * Side menu fragment
 */
public class SideMenuFragment extends Fragment {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = SideMenuFragment.class.getSimpleName();

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @BindView(R.id.drawer_header_user_name)
    TextView userName;

    @BindView(R.id.sidemenu_app_version)
    TextView appVersion;

    @BindView(R.id.sidemenu_app_debug_info)
    TextView appDebugInfo;

    @BindView(R.id.drawer_header_user_photo)
    ImageView userPhoto;

    @BindView(R.id.drawer_header_user_partner_logo)
    PartnerLogoImageView userPartnerLogo;

    @BindView(R.id.action_edit_profile)
    TextView userEditProfileTextView;

    @Nullable
    @BindView(R.id.action_update_info)
    SideMenuItemView updatePrivateCircleInfoItemView;

    @Nullable
    @BindView(R.id.action_charte)
    SideMenuItemView charterItemView;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------


    public SideMenuFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        BusProvider.INSTANCE.getInstance().register(this);
    }

    @Override
    public void onDestroy() {
        BusProvider.INSTANCE.getInstance().unregister(this);
        super.onDestroy();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.layout_side_menu, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initialiseView();
        updateUserView();
    }

    // ----------------------------------
    // BUS LISTENERS
    // ----------------------------------

    @Subscribe
    public void userInfoUpdated(Events.OnUserInfoUpdatedEvent event) {
        updateUserView();
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------

    private void initialiseView() {
        appVersion.setText(getString(R.string.about_version_format, BuildConfig.VERSION_FULL_NAME ));
        appDebugInfo.setText(getString(R.string.about_debug_info_format,BuildConfig.VERSION_DISPLAY_BRANCH_NAME, FirebaseInstanceId.getInstance().getId()));

        appVersion.setOnLongClickListener(v -> handleLongPress());
        appDebugInfo.setOnLongClickListener(v -> handleLongPress());

        //add listener to user photo and name, that opens the user profile screen
        userPhoto.setOnClickListener(v -> selectMenuAction(R.id.action_user));
        userName.setOnClickListener(v -> selectMenuAction(R.id.action_user));
        //add listener to modify profile text view
        userEditProfileTextView.setOnClickListener(v -> selectMenuAction(R.id.action_edit_profile));

        //add listeners to side menu items
        if (getView() != null) {
            LinearLayout sideMenuItemsLayout = getView().findViewById(R.id.sidemenuitems_layout);
            if (sideMenuItemsLayout != null) {
                int itemsCount = sideMenuItemsLayout.getChildCount();
                for (int j = 0; j < itemsCount; j++) {
                    View child = sideMenuItemsLayout.getChildAt(j);
                    if (child instanceof SideMenuItemView) {
                        child.setOnClickListener(v -> selectMenuAction(v.getId()));
                    }
                }
            }
        }
    }

    private void updateUserView() {
        User user = EntourageApplication.me(getActivity());
        if (user != null) {
            userName.setText(user.getDisplayName());
            String avatarURL = user.getAvatarURL();
            if (avatarURL != null) {
                Picasso.get()
                        .load(Uri.parse(avatarURL))
                        .placeholder(R.drawable.ic_user_photo_small)
                        .transform(new CropCircleTransformation())
                        .into(userPhoto);
            } else {
                userPhoto.setImageResource(R.drawable.ic_user_photo_small);
            }
            // Show partner logo
            String partnerURL = null;
            Partner partner = user.getPartner();
            if (partner != null) {
                partnerURL = partner.getSmallLogoUrl();
            }
            if (partnerURL != null) {
                Picasso.get()
                        .load(Uri.parse(partnerURL))
                        .placeholder(R.drawable.partner_placeholder)
                        .transform(new CropCircleTransformation())
                        .into(userPartnerLogo);
            } else {
                userPartnerLogo.setImageDrawable(null);
            }
            // Show Update Private Circle item only if the user is member of any
            if (updatePrivateCircleInfoItemView != null) {
                updatePrivateCircleInfoItemView.setVisibility( user.getMemberships(Entourage.TYPE_NEIGHBORHOOD).size() > 0 ? View.VISIBLE : View.GONE);
            }
            // Changed the ethics charter text depending on signed/unsigned
            if (charterItemView != null) {
                charterItemView.setTitle(user.hasSignedEthicsCharter() ? R.string.action_charter_signed : R.string.action_charter_unsigned);
            }
        }
    }

    private void selectMenuAction(int action) {
        if (getActivity() == null || !(getActivity() instanceof MainActivity)) return;
        MainActivity mainActivity = (MainActivity)getActivity();
        mainActivity.selectItem(action);
    }

    private boolean handleLongPress() {
        selectMenuAction(R.id.sidemenu_app_version);
        EntourageSnackbar.INSTANCE.make(
                getView().findViewById(R.id.sideMenuCoordinatorLayout),
                R.string.debug_info_clipboard,
                Snackbar.LENGTH_SHORT
        ).show();
        return true;
    }

}

File: app/src/main/java/social/entourage/android/sidemenu/SideMenuFragment.kt
package social.entourage.android.sidemenu

import android.net.Uri
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.iid.FirebaseInstanceId
import com.squareup.otto.Subscribe
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.entourage.layout_side_menu.*
import kotlinx.android.synthetic.entourage.layout_side_menu_appversion.*
import social.entourage.android.BuildConfig
import social.entourage.android.EntourageApplication
import social.entourage.android.MainActivity
import social.entourage.android.R
import social.entourage.android.api.tape.Events
import social.entourage.android.tools.BusProvider
import social.entourage.android.tools.CropCircleTransformation
import social.entourage.android.view.EntourageSnackbar

/**
 * Side menu fragment
 */
class SideMenuFragment  : Fragment(R.layout.layout_side_menu) {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        BusProvider.instance.register(this)
    }

    override fun onDestroy() {
        BusProvider.instance.unregister(this)
        super.onDestroy()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initialiseView()
        updateUserView()
    }

    // ----------------------------------
    // BUS LISTENERS
    // ----------------------------------
    @Subscribe
    fun userInfoUpdated(event: Events.OnUserInfoUpdatedEvent?) {
        updateUserView()
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------
    private fun initialiseView() {
        sidemenu_app_version?.text = getString(R.string.about_version_format, BuildConfig.VERSION_FULL_NAME)
        sidemenu_app_debug_info?.text = getString(R.string.about_debug_info_format, BuildConfig.VERSION_DISPLAY_BRANCH_NAME, FirebaseInstanceId.getInstance().id)
        sidemenu_app_version?.setOnLongClickListener { v: View? -> handleLongPress() }
        sidemenu_app_debug_info?.setOnLongClickListener { v: View? -> handleLongPress() }

        //add listener to user photo and name, that opens the user profile screen
        drawer_header_user_photo?.setOnClickListener { v: View? -> selectMenuAction(R.id.action_user) }
        drawer_header_user_name?.setOnClickListener { v: View? -> selectMenuAction(R.id.action_user) }
        //add listener to modify profile text view
        action_edit_profile?.setOnClickListener { v: View? -> selectMenuAction(R.id.action_edit_profile) }

        //add listeners to side menu items
        if (sidemenuitems_layout != null) {
            val itemsCount = sidemenuitems_layout.childCount
            for (j in 0 until itemsCount) {
                (sidemenuitems_layout.getChildAt(j) as? SideMenuItemView)?.setOnClickListener { v: View -> selectMenuAction(v.id) }
            }
        }
    }

    private fun updateUserView() {
        val user = EntourageApplication.me(activity) ?: return
        drawer_header_user_name?.text = user.displayName
        val avatarURL = user.avatarURL
        if (avatarURL != null && drawer_header_user_photo!=null) {
            Picasso.get()
                    .load(Uri.parse(avatarURL))
                    .placeholder(R.drawable.ic_user_photo_small)
                    .transform(CropCircleTransformation())
                    .into(drawer_header_user_photo!!)
        } else {
            drawer_header_user_photo?.setImageResource(R.drawable.ic_user_photo_small)
        }
        // Show partner logo
        var partnerURL: String? = null
        val partner = user.partner
        if (partner != null) {
            partnerURL = partner.smallLogoUrl
        }
        if (partnerURL != null && drawer_header_user_partner_logo!=null) {
            Picasso.get()
                    .load(Uri.parse(partnerURL))
                    .placeholder(R.drawable.partner_placeholder)
                    .transform(CropCircleTransformation())
                    .into(drawer_header_user_partner_logo!!)
        } else {
            drawer_header_user_partner_logo!!.setImageDrawable(null)
        }
        // Show Update Private Circle item only if the user is member of any
        //TODO ADD THIS TO PFP : action_update_info?.visibility = if (user.getMemberships(Entourage.TYPE_NEIGHBORHOOD).size > 0) View.VISIBLE else View.GONE
        // Changed the ethics charter text depending on signed/unsigned
        action_charte?.setTitle(if (user.hasSignedEthicsCharter()) R.string.action_charter_signed else R.string.action_charter_unsigned)
    }

    private fun selectMenuAction(action: Int) {
        if (activity == null || activity !is MainActivity) return
        (activity as MainActivity).selectItem(action)
    }

    private fun handleLongPress(): Boolean {
        selectMenuAction(R.id.sidemenu_app_version)
        if(sideMenuCoordinatorLayout != null) {
            EntourageSnackbar.make(
                    sideMenuCoordinatorLayout,
                    R.string.debug_info_clipboard,
                    Snackbar.LENGTH_SHORT
            ).show()
        }
        return true
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        val TAG = SideMenuFragment::class.java.simpleName
    }
}